text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,349 @@
+
+```yaml
+userMap:
+- originalName: yangkeao
+  cluster: *
+  name: admin
+```
+
+The webhook should also create a `SubjectAccessReview` in the remote cluster to
+check whether the remote user is privileged enough to create the chaos. This
+behavior makes it possible for the user to know the authorization error while
+creating the chaos, but not after creation.
+
+We should also provide an option to stop the webhook (in helm), as it will truly
+bring some inconvenience to the users. If the users don’t care about
+Authentication and Authorization, they can turn off this feature.
+
+### Q&A
+
+1. Why don’t we build a standalone authentication solution?
+
+   This sound never dimmed since the first day we were considering the
+   authentication. From the subjective aspect, as a developer, I don’t want to
+   build or ship a software with awful experience and strange configuration.
+   From the objective aspect, as a user, I don’t want to learn or manage the
+   authentication for every single application on my cluster. It would be a
+   disaster to create users / teams / groups in every application. Obviously,
+   the Chaos Mesh shouldn’t be the bad guy.
+
+   If there are some needs which cannot be solved through RBAC, I don’t regard
+   it as a common need (or a suitable need). e.g. you want to list every
+   namespaces which you have privilege…
+
+2. Why don't we store the user information inside the chaos?
+
+   Yes. We could store the username inside the chaos annotation, and during the
+   Reconcile, we impersonate the user. It is possible, but doesn’t solve the
+   problem better. With webhook, we can tell the user that you are not
+   privileged enough to create the resource at the creation time, which is
+   really expected (like every resource in a single cluster).
+
+3. What about the `Workflow` and `Schedule`?
+
+   The `Workflow` can create chaos in different clusters, and it’s the only way
+   to orchestrate the chaos in different clusters. We could get a list of
+   affected clusters/namespaces, and evaluate the privilege.
+
+4. What about creating a CRD for `userMap`?
+
+   I think a ConfigMap is enough. A standalone CRD cannot bring more benefits.
+
+5. `userMap` is ugly!
+
+   Yes. It reminds me of uid_map in the linux user_namespace, which is also an
+   ugly thing. If you have a better idea, please tell me.
+
+## Architecture
+
+The total architecture is really simple.
+
+In the management (or so-called parent) cluster, the chaos-controller-manager
+will be responsible for
+
+1. deploying chaos mesh to other clusters
+2. validating chaos
+3. authorization
+4. control workflow and schedule
+5. synchronizing the chaos status
+6. normal chaos injecting / recovering, as the user may want to inject to
+   current management cluster
+
+In the remote cluster, the chaos-controller-manager and chaos-daemon will be
+responsible for injecting / recovering chaos
+
+These two chaos-controller-manager could share a single binary, so the
```

Comment:
I don't think sharing a single binary is a good idea. The logic of parent mode` is different from `child mode`, if we share in a single binary, which will make chaos-controller-manager more complicated

Reply:
Well. It is not that different actually. Just like we can run `Schedule` and `Workflow` controller in a standalone binary, but running them in the same binary doesn't bring too much trouble.

At least, I found I'll need to clarify the statement. It seems that three modes are somehow duplicated, the remote cluster chaos mesh doesn't need to know itself is a child, and the parent chaos mesh should also be able to inject chaos into current (parent) cluster.